src/coreclr/jit/lsraxarch.cpp

```diff
@@ -1341,6 +1341,16 @@ int LinearScan::BuildCall(GenTreeCall* call)
         srcCount += BuildOperandUses(ctrlExpr, ctrlExprCandidates);
     }
 
+    if (call->NeedsVzeroupper(compiler))
```

Comment:
why this has to do in lsra and not somewhere else like when we create the `GT_CALL` ?

Reply:
Because its the place we currently do the same codegen or later tracking related to `SetContainsAVXFlags` and `SetContains256bitOrMoreAVX`.

This really comes down, in part, to register allocation related information. It's effectively part of the callee/caller save state information. The call means LSRA must save all the upper bits of YMM/ZMM registers and additionally for performance reasons, we are explicitly zeroing the upper registers.

Reply:
> Because its the place we currently do the same codegen or later tracking related to SetContainsAVXFlags and SetContains256bitOrMoreAVX.

The information we are tracking is not dependent on any of the lsra state nor is it consumed by lsra in anyway. It is purely a codegen information that IMO should be tracked as part of `genCodeForTreeNode()` or equivalent e.g. `genCodeForStoreBlk()` or `genCall()`, etc.

Reply:
It cannot be done in `genCall` as it is needed prior to the individual call node having code generated.

It's a heuristic over the entirety of the method in the exact same way that `SetContains256bitOrMoreAVX` is. Both `Contains256bitOrMoreAVX` and `ContainsCallNeedingVzeroupper` are then used in coordination with each other.

`SetContains256bitOrMoreAVX` was placed in LSRA for much the same reasons. It is a heuristic over the entire method that is needed before all nodes for the method have undergone codegen. It needs to be instanatiated late based on what nodes still persist. If it was done prior to rationalization or lowering, then it may miss nodes that are introduced including helper calls or transformations of nodes into hardware intrinsics that then utilize YMM or ZMM state.

Reply:
`Contains256bitOrMoreAVX ` and `ContainsCallNeedingVzeroupper` are used during creation of prologs/epilogs which happen after we generate code for all the nodes, so technically setting those bits can be done in codegen unless I am missing something here. 

 https://github.com/dotnet/runtime/blob/443ece48d04eff6f09c9ab34e45fbb20b457f4e8/src/coreclr/jit/codegencommon.cpp#L1923-L1936

Reply:
They are also used when generating calls themselves in which case we need to know if any other node has used 256-bit or higher AVX and if any call potentially needs vzeroupper.

This isn't information that can be done in codegen, it's why `Contains256bitOrMoreAVX` was placed here in LSRA 7 years ago and similarly why I've put `ContainsCallNeedingVzeroupper` here alongside that.